src/lib.rs

```diff
@@ -3285,6 +3290,66 @@ impl Build {
         Ok(ret)
     }
 
+    fn apple_deployment_version(
+        &self,
+        os: AppleOs,
+        target: &str,
+        arch_str: Option<&str>,
+    ) -> String {
+        fn rustc_provided_target(rustc: Option<&str>, target: &str) -> Option<String> {
+            let rustc = rustc?;
+            let output = Command::new(rustc)
+                .stderr(Stdio::piped())
+                .stdout(Stdio::piped())
+                .args(&["--target", target])
+                .args(&["--print", "deployment-target"])
+                .output()
+                .ok()?;
+
+            if output.status.success() {
+                std::str::from_utf8(&output.stdout)
+                    .unwrap()
+                    .split('=')
+                    .last()
+                    .map(|v| v.trim())
+                    .map(ToString::to_string)
+            } else {
+                // rustc is < 1.71
+                None
+            }
+        }
+
+        let rustc = self.getenv("RUSTC");
+        let rustc = rustc.as_deref();
+        // note the hardcoded minimums here are subject to change in a future compiler release,
+        // so the ones rustc knows about are preferred. For any compiler version that has bumped them
+        // though, `--print deployment-target` will be present and the fallbacks won't be used.
+        //
+        // the ordering of env -> rustc -> old defaults is intentional for performance when using
+        // an explicit target
+        match os {
+            AppleOs::MacOs => env::var("MACOSX_DEPLOYMENT_TARGET")
```

Comment:
Hm... should we emit rerun-if-changed flags for these env vars in any situation? (Perhaps it would be needed in complex situations with cross-compiling?)

Reply:
Originally I opted to avoid that (a prior revision of the branch actually ran it through the caching env paths) for consistency reasons. Only `cc` would rerun, so an arbitrary number of objects not created by `cc` in the build tree would have a different deployment target and that seems worse then doing nothing and making people run `cargo clean` to switch.

I think that if this is desired behavior, `rustc` or `cargo` should handle it instead and invalidate the entire build tree at once.